File: lib/sentinel/policy.ts
import type { PolicyDecision, RemediationAction, RiskLevel } from "./types";

const ACTION_POLICY: Record<RemediationAction, { risk: RiskLevel; permission: string; productionApproval: boolean; alwaysApproval?: boolean }> = {
  no_action: { risk: 0, permission: "READ_INCIDENT", productionApproval: false },
  send_notification: { risk: 1, permission: "SEND_NOTIFICATION", productionApproval: false },
  create_ticket: { risk: 1, permission: "CREATE_TICKET", productionApproval: false },
  refresh_key_cache: { risk: 1, permission: "REFRESH_CACHE", productionApproval: false },
  increase_retry_backoff: { risk: 1, permission: "UPDATE_RETRY_POLICY", productionApproval: true },
  pause_consumer: { risk: 2, permission: "OPERATE_CONSUMER", productionApproval: true },
  restart_service: { risk: 2, permission: "SERVICE_RESTART", productionApproval: true },
  scale_service: { risk: 2, permission: "SERVICE_SCALE", productionApproval: true },
  rollback_deployment: { risk: 3, permission: "DEPLOYMENT_ROLLBACK", productionApproval: true, alwaysApproval: true },
};

export function evaluatePolicy(input: {
  action: RemediationAction;
  environment: "production" | "staging";
  role?: "viewer" | "engineer" | "incident_commander" | "senior_engineer";
}): PolicyDecision {
  const role = input.role ?? "incident_commander";
  const config = ACTION_POLICY[input.action];
  const allowedRoles = config.risk >= 3 ? ["incident_commander", "senior_engineer"] : ["engineer", "incident_commander", "senior_engineer"];
  const allowed = input.action === "no_action" || allowedRoles.includes(role);
  const requiresApproval = allowed && (config.alwaysApproval === true || (input.environment === "production" && config.productionApproval));

  return {
    action: input.action,
    risk: config.risk,
    allowed,
    requiresApproval,
    permission: config.permission,
    reason: !allowed
      ? `${role} is not permitted to request ${input.action}.`
      : requiresApproval
        ? `${input.action} is a risk level ${config.risk} action in ${input.environment}; deterministic policy requires a human approval.`
        : `${input.action} is within the current role and environment policy boundary.`,
  };
}

export function isKnownAction(action: string): action is RemediationAction {
  return action in ACTION_POLICY;
}

File: sentinel/backend/app/policy.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum


class RiskLevel(IntEnum):
    INFORMATIONAL = 0
    REVERSIBLE = 1
    PRODUCTION_OPERATIONAL = 2
    DESTRUCTIVE = 3


@dataclass(frozen=True)
class ActionPolicy:
    permission: str
    risk: RiskLevel
    production_approval: bool = False
    always_approval: bool = False


@dataclass(frozen=True)
class PolicyDecision:
    action: str
    permission: str
    risk: RiskLevel
    allowed: bool
    requires_approval: bool
    reason: str


POLICIES: dict[str, ActionPolicy] = {
    "no_action": ActionPolicy("READ_INCIDENT", RiskLevel.INFORMATIONAL),
    "send_notification": ActionPolicy("SEND_NOTIFICATION", RiskLevel.REVERSIBLE),
    "create_ticket": ActionPolicy("CREATE_TICKET", RiskLevel.REVERSIBLE),
    "refresh_key_cache": ActionPolicy("REFRESH_CACHE", RiskLevel.REVERSIBLE),
    "increase_retry_backoff": ActionPolicy("UPDATE_RETRY_POLICY", RiskLevel.REVERSIBLE, production_approval=True),
    "pause_consumer": ActionPolicy("OPERATE_CONSUMER", RiskLevel.PRODUCTION_OPERATIONAL, production_approval=True),
    "restart_service": ActionPolicy("SERVICE_RESTART", RiskLevel.PRODUCTION_OPERATIONAL, production_approval=True),
    "scale_service": ActionPolicy("SERVICE_SCALE", RiskLevel.PRODUCTION_OPERATIONAL, production_approval=True),
    "rollback_deployment": ActionPolicy("DEPLOYMENT_ROLLBACK", RiskLevel.DESTRUCTIVE, production_approval=True, always_approval=True),
}


def evaluate_action(action: str, environment: str, role: str) -> PolicyDecision:
    policy = POLICIES.get(action)
    if policy is None:
        return PolicyDecision(action, "UNKNOWN", RiskLevel.DESTRUCTIVE, False, False, "Unknown action is denied by default.")

    if action == "no_action":
        allowed = True
    elif policy.risk >= RiskLevel.DESTRUCTIVE:
        allowed = role in {"incident_commander", "senior_engineer"}
    else:
        allowed = role in {"engineer", "incident_commander", "senior_engineer"}

    requires_approval = allowed and (
        policy.always_approval
        or (environment == "production" and policy.production_approval)
    )

    if not allowed:
        reason = f"Role {role!r} lacks {policy.permission}."
    elif requires_approval:
        reason = f"Risk level {int(policy.risk)} action in {environment} requires a human approval."
    else:
        reason = "Action is inside the current deterministic policy boundary."

    return PolicyDecision(
        action=action,
        permission=policy.permission,
        risk=policy.risk,
        allowed=allowed,
        requires_approval=requires_approval,
        reason=reason,
    )
